index.src.html

```diff
@@ -185,17 +190,43 @@
     When invoked, run these steps:
     1. If |config| is not a <a>valid AudioDecoderConfig</a>, throw a
         {{TypeError}}.
-    2. Run the <a>Configure Decoder</a> algorithm with |config|.
+    2. If {{AudioDecoder/state}} is `“closed”`, throw an {{InvalidStateError}}.
+    3. If the user agent cannot provide a codec implementation to support
+        config, throw a {{NotSupportedError}}.
+    4. Set {{AudioDecoder/state}} to `"configured"`.
+    5. <a>Queue a control message</a> to configure the decoder with |config|.
+    6. <a>Run the control message processing loop</a>.
```

Comment:
> We have a control thread. It can be a web worker or the main thread (for now). Methods are called on objects. Those methods have a synchronous part and/or an asynchronous part.
> 
> It seems that the general patter is to do some synchronous steps, and then queue a control message, and then pumping synchronously this new queue, this performs some steps, or returns. If the control message hasn't been executed (early return), it needs to be executed some other time, but we don't know when how or where. If the control message has been executed then something happens and then the rest of the steps are going to happen automatically.

This is a good summary of the current design. I didn't see anything prescriptive in the summary though - lmk if I'm missing a suggestion. 

> In addition, it seems that there is nothing in the synchronous part of the control messages (again, for now), that requires synchronicity, but maybe you have something in mind of the future?

Maybe the one thing that is required to be synchronous is modification/reading of `codecState`. For example, it is an error to call `decode()` without first calling `configure()`. This is verified by `configure()` synchronously setting `codecState` to `configured` and having `decode()` synchronously check that. Maybe we could make both of these asynchronous as long as we kept them in sequence. Would that be an improvement (unlock simpler language)? One nicety of the current design is its easy for the user to reason about codecState (which is exposed). Maybe we could still go async and fire events as codecState updates (e.g. <video>.readyState). 

Worth mentioning: `reset()` is entirely synchronous. In practice, there are some async steps still under the hood, but we definitely take care to sychronously disconnect the codec output callback from the underlying implementation.

Reply:
> We have a control thread. It can be a web worker or the main thread (for now). Methods are called on objects. Those methods have a synchronous part and/or an asynchronous part.
> 
> It seems that the general patter is to do some synchronous steps, and then queue a control message, and then pumping synchronously this new queue, this performs some steps, or returns. If the control message hasn't been executed (early return), it needs to be executed some other time, but we don't know when how or where. If the control message has been executed then something happens and then the rest of the steps are going to happen automatically.

This is a good summary of the current design. I didn't see anything prescriptive in the summary though - lmk if I'm missing a suggestion. 

> In addition, it seems that there is nothing in the synchronous part of the control messages (again, for now), that requires synchronicity, but maybe you have something in mind of the future?

Maybe the one thing that is required to be synchronous is modification/reading of `codecState`. For example, it is an error to call `decode()` without first calling `configure()`. This is verified by `configure()` synchronously setting `codecState` to `configured` and having `decode()` synchronously check that. Maybe we could make both of these asynchronous as long as we kept them in sequence. Would that be an improvement (unlock simpler language)? One nicety of the current design is its easy for the user to reason about codecState (which is exposed). Maybe we could still go async and fire events as codecState updates (e.g. <video>.readyState). 

Worth mentioning: `reset()` is entirely synchronous. In practice, there are some async steps still under the hood, but we definitely take care to sychronously disconnect the codec output callback from the underlying implementation.

Reply:
Over VC we discussed how the UA restarts the loop at its discretion a little further down. I agreed to change the language to something like: "when UA determines the blockage is resolved, post a task to the ??? task source main thread to resume the Media processing loop"

I haven't made that update yet. Waiting for our discussion of queues to resolve.

Reply:
Resolving: obsolete with latest commit to define explicit threads.